lib.rs

```diff
@@ -2,45 +2,126 @@
+            match err {
+                MerkleTreeError::MerkleTreeIsFull => return Error::MerkleTreeIsFull,
+                MerkleTreeError::DepthTooLong => return Error::MerkleTreeInvalidDepth,
+                MerkleTreeError::DepthIsZero => return Error::MerkleTreeInvalidDepth, ////
+            }
         }
 
-        /// Constructor that initializes the `bool` value to `false`.
+    }
+
+    pub type Result<T> = core::result::Result<T, Error>;
+
+    impl Slushie {
+        /// create a new Slushie contract
         ///
-        /// Constructors can delegate to other constructors.
+        /// Takes the deposit_size Balance amount
+        /// so the users can deposit and withdraw
+        /// only a fixed amount of tokens.
         #[ink(constructor)]
-        pub fn default() -> Self {
-            Self::new(Default::default())
+        pub fn new(deposit_size: Balance) -> Self {
+            ink::utils::initialize_contract(|me: &mut Self| {
+                *me = Self { merkle_tree: MerkleTree::<MAX_DEPTH, DEFAULT_ROOT_HISTORY_SIZE>::new().unwrap(),
+                    deposit_size,
+                    used_nullifiers: Default::default(),
+                };
+            })
         }
 
-        /// A message that can be called on instantiated contracts.
-        /// This one flips the value of the stored `bool` from `true`
-        /// to `false` and vice versa.
-        #[ink(message)]
-        pub fn flip(&mut self) {
-            self.value = !self.value;
+        #[ink(message, payable)]
+        pub fn deposit(&mut self, hash: PoseidonHash) -> Result<PoseidonHash> {
+            self.merkle_tree.insert(hash)?;
+
+            if self.env().transferred_value() != self.deposit_size {
+                return Err(Error::InvalidTransferredAmount); // FIXME: suggest a better name
+            }
+
+            self.env().emit_event(
+                DepositEvent {
+                    hash,
+                    timestamp: self.env().block_timestamp(),
+                });
+
+            Ok(self.merkle_tree.get_last_root() as PoseidonHash)
         }
 
-        /// Simply returns the current value of our `bool`.
         #[ink(message)]
-        pub fn get(&self) -> bool {
-            self.value
+        pub fn withdraw(&mut self, hash: PoseidonHash, root: PoseidonHash) -> Result<()> {
+            if !self.merkle_tree.is_known_root(root) {
+                return Err(Error::UnknownRoot);
+            }
+
+            if self.env().balance() < self.deposit_size {
+                return Err(Error::WithdrawalFailureInsufficientFunds);
+            }
+
+            if self.env().transfer(self.env().caller(), self.deposit_size).is_err() {
+                return Err(Error::WithdrawalFailureInvalidDepositSize);
+            }
+
+            if self.used_nullifiers.get(hash).is_some() {
+                return Err(Error::WithdrawalFailureNullifierAlreadyUsed);
+            }
+
+            self.used_nullifiers.insert(hash, &true);
```

Comment:
```suggestion
            self.used_nullifiers.insert(hash, true);
```
I think it should work fine without the reference

Reply:
Actually not, it was used in the first place, but produced compilation error.


`161 |             self.used_nullifiers.insert(hash, true);
    |                                               ^^^^
    |                                               |
    |                                               expected reference, found `bool`
    |                                               help: consider borrowing here: `&true`
    |
    = note: expected reference `&_`
                    found type `bool`
`